kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/ProxyChannelStateMachine.java

```diff
@@ -0,0 +1,532 @@
+     */
+    @SuppressWarnings({ "DataFlowIssue", "java:S2637" })
+    @NonNull
+    private KafkaProxyFrontendHandler frontendHandler = null;
+
+    /**
+     * The backend handler. Non-null if {@link #onNetFilterInitiateConnect(HostPort, List, VirtualCluster, NetFilter)}
+     * has been called
+     */
+    @VisibleForTesting
+    @Nullable
+    private KafkaProxyBackendHandler backendHandler;
+
+    ProxyChannelState state() {
+        return state;
+    }
+
+    /**
+     * Purely for tests DO NOT USE IN PRODUCTION code!!
+     * Sonar will complain if one uses this in prod code listen to it.
+     */
+    @VisibleForTesting
+    void forceState(@NonNull ProxyChannelState state, @NonNull KafkaProxyFrontendHandler frontendHandler, @Nullable KafkaProxyBackendHandler backendHandler) {
+        LOGGER.info("Forcing state to {} with {} and {}", state, frontendHandler, backendHandler);
+        this.state = state;
+        this.frontendHandler = frontendHandler;
+        this.backendHandler = backendHandler;
+    }
+
+    @Override
+    public String toString() {
+        return "StateHolder{" +
+                "state=" + state +
+                ", serverReadsBlocked=" + serverReadsBlocked +
+                ", clientReadsBlocked=" + clientReadsBlocked +
+                ", frontendHandler=" + frontendHandler +
+                ", backendHandler=" + backendHandler +
+                '}';
+    }
+
+    public String currentState() {
+        return this.state().getClass().getSimpleName();
+    }
+
+    /**
+     * Notify the state machine when the client applies back pressure.
+     */
+    public void onClientUnwritable() {
+        if (!serverReadsBlocked) {
+            serverReadsBlocked = true;
+            Objects.requireNonNull(backendHandler).applyBackpressure();
+        }
+    }
+
+    /**
+     * Notify the state machine when the client stops applying back pressure
+     */
+    public void onClientWritable() {
+        if (serverReadsBlocked) {
+            serverReadsBlocked = false;
+            Objects.requireNonNull(backendHandler).relieveBackpressure();
+        }
+    }
+
+    /**
+     * Notify the state machine when the server applies back pressure
+     */
+    public void onServerUnwritable() {
+        if (!clientReadsBlocked) {
+            clientReadsBlocked = true;
+            frontendHandler.applyBackpressure();
+        }
+    }
+
+    /**
+     * Notify the state machine when the server stops applying back pressure
+     */
+    public void onServerWritable() {
+        if (clientReadsBlocked) {
+            clientReadsBlocked = false;
+            frontendHandler.relieveBackpressure();
+        }
+    }
+
+    /**
+     * Notify the statemachine that the client channel has an active TCP connection.
+     * @param frontendHandler with active connection
+     */
+    void onClientActive(@NonNull KafkaProxyFrontendHandler frontendHandler) {
+        if (STARTING_STATE.equals(this.state)) {
```

Comment:
I'm wondering a little about whether you could potentially hide the different state implementations from this machine. I wonder what it would look like if we had all these `on*` methods also defined on ProxyChannelState and returning `Optional<ProxyChannelState>` with a default behaviour of throwing an IllegalStateException.

Then instead of having instanceOfs to switch types you'd forward the `on*` event through and if you get a completed Optional back you know a state transition has been made. Though I'm not sure how you'd wire up the feedback to the handlers to call the relevant `in*` methods, maybe it could be a job for the state.

Keen to have a play after this is merged anyway. Sealed interface with a switch is pretty cool as it is, and maybe it's good to trap the Handler call back responsibility in this one class.

Reply:
Not quite sure what you mean by a completed optional. Do you mean one where the new state is present? 

I think that could be interesting to look at certainly.